components/webext-storage/src/sync/outgoing.rs

```diff
@@ -0,0 +1,185 @@
+               LEFT JOIN storage_sync_mirror m ON m.ext_id = l.ext_id
+               LEFT JOIN storage_sync_staging s ON s.ext_id = l.ext_id
+               WHERE sync_change_counter > 0";
+    let elts = tx
+        .conn()
+        .query_rows_and_then_named(sql, &[], outgoing_from_row)?;
+
+    log::debug!("get_outgoing found {} items", elts.len());
+    // Now the temp table thang...
+    let tt_sql = "INSERT INTO storage_sync_outgoing_staging
+                  (guid, ext_id, data, sync_change_counter)
+                  VALUES (:guid, :ext_id, :data, :sync_change_counter)";
+    for (payload, change_counter) in &elts {
+        log::trace!("outgoing '{:?}' with counter={}", payload, change_counter);
+        tx.execute_named_cached(
+            tt_sql,
+            rusqlite::named_params! {
+                ":guid": payload.guid,
+                ":ext_id": payload.ext_id,
+                ":data": payload.data,
+                ":sync_change_counter": change_counter
+            },
+        )?;
+    }
+    Ok(elts.into_iter().map(|e| e.0).collect())
+}
+
+/// Record the fact that items were uploaded. This updates the state of the
+/// local DB to reflect the state of the server we just updated.
+/// Note that this call is almost certainly going to be made in a *different*
+/// transaction than the transaction used in `get_and_record_outgoing()`
+pub fn record_uploaded(
+    tx: &Transaction<'_>,
+    items: &[SyncGuid],
+    signal: &dyn Interruptee,
+) -> Result<()> {
+    log::debug!(
+        "record_uploaded recording that {} items were uploaded",
+        items.len()
+    );
+
+    let sql = "
+        UPDATE storage_sync_data SET
+            sync_change_counter =
+                (sync_change_counter - (
+                    SELECT sync_change_counter
+                    FROM storage_sync_outgoing_staging
+                    WHERE storage_sync_outgoing_staging.guid = :guid)
+                )
+        WHERE ext_id = (SELECT ext_id
+                        FROM storage_sync_outgoing_staging
+                        WHERE storage_sync_outgoing_staging.guid = :guid)";
+    for guid in items.iter() {
+        signal.err_if_interrupted()?;
+        log::trace!("recording guid='{}' was uploaded", guid);
+        tx.execute_named(
+            sql,
+            rusqlite::named_params! {
+                ":guid": guid,
+            },
+        )?;
+    }
+
+    // Copy incoming staging into the mirror, then outgoing staging to the
+    // mirror and local tombstones.
+    tx.execute_batch(
+        "
+        INSERT OR REPLACE INTO storage_sync_mirror (guid, ext_id, data)
+        SELECT guid, ext_id, data FROM temp.storage_sync_staging;
+
+        INSERT OR REPLACE INTO storage_sync_mirror (guid, ext_id, data)
+        SELECT guid, ext_id, data FROM temp.storage_sync_outgoing_staging;
+
+        -- XXX - is this correct?
+        DELETE from storage_sync_data WHERE data IS NULL AND sync_change_counter = 0;",
```

Comment:
I think so—durable tombstones get stored in the mirror, so there's no need for them to hang around locally once we've synced them.